crates/telio-wg/src/pq.rs

```diff
@@ -0,0 +1,351 @@
+/// a WG preshared key for subsequent connection
+pub async fn fetch_keys(
+    endpoint: impl ToSocketAddrs,
+    secret: &telio_crypto::SecretKey,
+    peers_pubkey: &telio_crypto::PublicKey,
+) -> Result<PqKeys, Error> {
+    let (tunn, sock) = handshake(endpoint, secret, peers_pubkey).await?;
+
+    let mut rng = rand::rngs::StdRng::from_entropy();
+
+    // Generate keys
+    let wg_secret = x25519_dalek::StaticSecret::random_from_rng(&mut rng);
+    let (pq_public, pq_secret) = kyber768::keypair();
+    let wg_public = x25519_dalek::PublicKey::from(&wg_secret);
+
+    // Send GET packet
+    let pkgbuf = create_get_packet(&wg_public, &pq_public, &mut rng); // 4 KiB
+
+    let mut recvbuf = [0u8; 2048]; // 2 KiB buffer should suffice
+    match tunn.encapsulate(&pkgbuf, &mut recvbuf) {
+        noise::TunnResult::Err(err) => {
+            return Err(format!("Failed to encapsulate PQ keys message: {err:?}").into())
+        }
+        noise::TunnResult::WriteToNetwork(buf) => {
+            sock.send(buf).await?;
+        }
+        _ => return Err("Unexpected WG tunnel output".into()),
+    }
+
+    // Receive response
+    let read = sock.recv(&mut recvbuf).await?;
+
+    let mut msgbuf = [0u8; 2048]; // 2 KiB buffer should shuffice
+
+    #[allow(index_access_check)]
+    let ciphertext = match tunn.decapsulate(None, &recvbuf[..read], &mut msgbuf) {
+        noise::TunnResult::Err(err) => {
+            return Err(format!("Failed to decapsulate PQ keys message: {err:?}").into())
+        }
+        noise::TunnResult::WriteToTunnelV4(buf, _) => parse_get_response(buf)?,
+        _ => return Err("Unexpected WG tunnel output".into()),
+    };
+
+    // Extract the shared secret
+    let pq_shared = kyber768::decapsulate(&ciphertext, &pq_secret);
+
+    Ok(PqKeys {
+        pq_shared,
+        wg_secret,
+    })
+}
+
+async fn handshake(
+    endpoint: impl ToSocketAddrs,
+    secret: &telio_crypto::SecretKey,
+    peers_pubkey: &telio_crypto::PublicKey,
+) -> Result<(Box<noise::Tunn>, UdpSocket), Error> {
+    let sock = UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0)).await?;
+    sock.connect(endpoint).await?;
+
+    let tunn = noise::Tunn::new(
+        secret.into_bytes().into(),
+        peers_pubkey.0.into(),
+        None,
+        None,
+        0,
+        None,
+    )?;
+
+    let mut pkgbuf = [0u8; 2048];
+    match tunn.encapsulate(&[], &mut pkgbuf) {
+        noise::TunnResult::Err(err) => {
+            return Err(format!("Failed to encapsulate handshake message: {err:?}").into())
+        }
+        noise::TunnResult::WriteToNetwork(buf) => {
+            sock.send(buf).await?;
+        }
+        _ => return Err("Unexpected WG tunnel output".into()),
+    }
+
+    // The response should be 92, so the buffer is sufficient
+    let read = sock.recv(&mut pkgbuf).await?;
+
+    let mut msgbuf = [0u8; 2048];
+
+    #[allow(index_access_check)]
+    match tunn.decapsulate(None, &pkgbuf[..read], &mut msgbuf) {
```

Comment:
Please do not just `#[allow(index_access_check)]` as if this got merged, it would introduce Denial-Of-Service vulnerability in networks with higher MTU than 2048.

Reply:
The same comment applies https://github.com/NordSecurity/libtelio/pull/247#discussion_r1431019907